ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -88,32 +103,41 @@ impl ProofBatch {
         let (mut uv_values, first_proof_from_left, my_first_proof_left_share) =
             LargeProofGenerator::gen_artefacts_from_recursive_step(
                 ctx,
-                &mut record_counter,
+                &mut prss_record_ids,
                 &first_lagrange_table,
                 ProofBatch::polynomials_from_inputs(uv_tuple_inputs),
             );
 
-        // approximate length of proof vector (rounded up)
-        let uv_len_bits: u32 = usize::BITS - uv_values.len().leading_zeros();
-        let small_recursion_factor_bits: u32 = usize::BITS - SRF.leading_zeros();
-        let expected_len = 1 << (uv_len_bits - small_recursion_factor_bits);
+        // `MAX_PROOF_RECURSION - 2` because:
+        //  * The first level of recursion has already happened.
+        //  * We need (SRF - 1) at the last level to have room for the masks.
+        let max_uv_values: usize =
+            (SRF - 1) * SRF.pow(u32::try_from(MAX_PROOF_RECURSION - 2).unwrap());
+        tracing::info!("uv_values.len() = {}", uv_values.len());
+        assert!(
+            uv_values.len() <= max_uv_values,
+            "Proof batch is too large: have {} uv_values, max is {}",
+            uv_values.len(),
+            max_uv_values,
+        );
 
         // storage for other proofs
-        let mut my_proofs_left_shares = Vec::<[Fp61BitPrime; SPL]>::with_capacity(expected_len);
+        let mut my_proofs_left_shares =
+            Vec::<[Fp61BitPrime; SPL]>::with_capacity(MAX_PROOF_RECURSION - 1);
         let mut shares_of_proofs_from_prover_left =
-            Vec::<[Fp61BitPrime; SPL]>::with_capacity(expected_len);
+            Vec::<[Fp61BitPrime; SPL]>::with_capacity(MAX_PROOF_RECURSION - 1);
 
         // generate masks
         // Prover `P_i` and verifier `P_{i-1}` both compute p(x)
         // therefore the "right" share computed by this verifier corresponds to that which
         // was used by the prover to the right.
-        let (my_p_mask, p_mask_from_right_prover) = ctx.prss().generate_fields(record_counter);
-        record_counter += 1;
+        let (my_p_mask, p_mask_from_right_prover) =
+            ctx.prss().generate_fields(prss_record_ids.next().unwrap());
         // Prover `P_i` and verifier `P_{i+1}` both compute q(x)
         // therefore the "left" share computed by this verifier corresponds to that which
         // was used by the prover to the left.
-        let (q_mask_from_left_prover, my_q_mask) = ctx.prss().generate_fields(record_counter);
-        record_counter += 1;
+        let (q_mask_from_left_prover, my_q_mask) =
```

Comment:
`record_id_range.next().unwrap()` is dominantly used everywhere it seems. I think we should either implement a helper function to make it look less ugly at callsite and potentially not implement the iterator trait for it as we don't have (unless I missed it) `for x in range` use case

Reply:
Thoughts on what to call the helper? Maybe `expect_next()`?